Control/Concurrent/MVar/Strict.hs

```diff
@@ -88,68 +96,31 @@ newMVar value =
     putMVar mvar value  >>
     return mvar
 
-{-|
-  This is a combination of 'takeMVar' and 'putMVar'; ie. it takes the value
-  from the 'MVar', puts it back, and also returns it.
--}
-readMVar :: NFData a => MVar a -> IO a
-readMVar m = mask $ \_ -> do
-    a <- takeMVar m
-    putMVar m a
-    return a
-
 {-|
   Take a value from an 'MVar', put a new value into the 'MVar' and
   return the value taken. Note that there is a race condition whereby
   another process can put something in the 'MVar' after the take
   happens but before the put does.
 -}
 swapMVar :: NFData a => MVar a -> a -> IO a
-swapMVar mvar new = mask $ \_ -> do
+swapMVar !mvar new = seq (force new) $ mask $ \_ -> do
```

Comment:
Shouldn't the local `putMVar` suffice, like the original version?

Reply:
I believe we want to force manually *before* we take the `MVar`. That way we don't hold the `MVar` any longer than necessary. We also avoid problems if the forcing throws an exception. Once we've forced manually, we don't want to do it again implicitly, because that would be a waste of time.